### PR TITLE
Immutable HeaderValue objects.

### PR DESCRIPTION
**Pull request by Tratcher**
#177 
@muratg @rynowak

### CODE REVIEW

File: src/Microsoft.Net.Http.Headers/MediaTypeHeaderValue.cs [hunk 1]
```diff
@@ -20,8 +21,9 @@ public class MediaTypeHeaderValue
             = new GenericHeaderParser<MediaTypeHeaderValue>(true, GetMediaTypeLength);
 
         // Use list instead of dictionary since we may have multiple parameters with the same name.
```
Comment (rynowak) [inline src/Microsoft.Net.Http.Headers/MediaTypeHeaderValue.cs]: Comment is stale?

File: src/Microsoft.Net.Http.Headers/MediaTypeHeaderValue.cs [hunk 2]
```diff
@@ -250,22 +268,47 @@ public bool IsSubsetOf(MediaTypeHeaderValue otherMediaType)
 
         /// <summary>
         /// Performs a deep copy of this object and all of it's NameValueHeaderValue sub components,
-        /// while avoiding the cost of revalidating the components.
+        /// while avoiding the cost of re-validating the components.
```
Comment (rynowak) [inline src/Microsoft.Net.Http.Headers/MediaTypeHeaderValue.cs]: it was better before

File: src/Microsoft.Net.Http.Headers/MediaTypeHeaderValue.cs [hunk 3]
```diff
@@ -250,22 +268,47 @@ public bool IsSubsetOf(MediaTypeHeaderValue otherMediaType)
...
+            }
+            return other;
+        }
+
+        /// <summary>
+        /// Performs a deep copy of this object and all of it's NameValueHeaderValue sub components,
+        /// while avoiding the cost of re-validating the components. This copy is read-only.
+        /// </summary>
+        /// <returns>A deep, read-only, copy.</returns>
+        public MediaTypeHeaderValue CopyAsReadOnly()
```
Comment (muratg) [inline src/Microsoft.Net.Http.Headers/MediaTypeHeaderValue.cs]: Perhaps `CloneAsReadOnly`?

**Comment by rynowak** [comment]
:shipit: from me, seems like it would solve the issue we have.

**Comment by muratg** [comment]
:shipit:

**Pull request merged**